homeassistant/components/opentherm_gw/services.py

```diff
@@ -0,0 +1,296 @@
+"""Support for OpenTherm Gateway devices."""
+
+from __future__ import annotations
+
+from datetime import date, datetime
+from typing import TYPE_CHECKING
+
+import pyotgw.vars as gw_vars
+import voluptuous as vol
+
+from homeassistant.const import (
+    ATTR_DATE,
+    ATTR_ID,
+    ATTR_MODE,
+    ATTR_TEMPERATURE,
+    ATTR_TIME,
+)
+from homeassistant.core import HomeAssistant, ServiceCall
+from homeassistant.exceptions import ServiceValidationError
+from homeassistant.helpers import config_validation as cv
+
+from .const import (
+    ATTR_CH_OVRD,
+    ATTR_DHW_OVRD,
+    ATTR_GW_ID,
+    ATTR_LEVEL,
+    ATTR_TRANSP_ARG,
+    ATTR_TRANSP_CMD,
+    DATA_GATEWAYS,
+    DATA_OPENTHERM_GW,
+    DOMAIN,
+    SERVICE_RESET_GATEWAY,
+    SERVICE_SEND_TRANSP_CMD,
+    SERVICE_SET_CH_OVRD,
+    SERVICE_SET_CLOCK,
+    SERVICE_SET_CONTROL_SETPOINT,
+    SERVICE_SET_GPIO_MODE,
+    SERVICE_SET_HOT_WATER_OVRD,
+    SERVICE_SET_HOT_WATER_SETPOINT,
+    SERVICE_SET_LED_MODE,
+    SERVICE_SET_MAX_MOD,
+    SERVICE_SET_OAT,
+    SERVICE_SET_SB_TEMP,
+)
+
+if TYPE_CHECKING:
+    from . import OpenThermGatewayHub
+
+
+def _get_gateway(call: ServiceCall) -> OpenThermGatewayHub:
+    gw_id: str = call.data[ATTR_GW_ID]
+    gw_hub: OpenThermGatewayHub | None = (
+        call.hass.data.get(DATA_OPENTHERM_GW, {}).get(DATA_GATEWAYS, {}).get(gw_id)
+    )
+    if gw_hub is None:
+        raise ServiceValidationError(
+            translation_domain=DOMAIN,
+            translation_key="invalid_gateway_id",
+            translation_placeholders={"gw_id": gw_id},
+        )
+    return gw_hub
+
+
+def register_services(hass: HomeAssistant) -> None:
+    """Register services for the component."""
+    service_reset_schema = vol.Schema({vol.Required(ATTR_GW_ID): vol.All(cv.string)})
```

Comment:
Note: the schemas and the functions could now be moved to be module-level constants and functions, but I plan to do that in a follow-up PR to make it easier to review